docs/design_proposals/artifact-dependencies.md

```diff
@@ -0,0 +1,223 @@
+# Supporting build artifact dependencies
+
+## Background
+Refer [this](https://tinyurl.com/skaffold-modules) document which presents the design goals for introducing the concept of _modules_ in Skaffold. A prerequisite to being able to define modules and supporting cross module dependency is to first consider the latest version of the skaffold config (_currently `v2beta8`_) as an implicit module and allowing dependencies between artifacts defined in it. The current document aims to capture the major code changes necessary for achieving this.
+
+>*Note: Omitted some details for brevity, assuming reader's familiarity with the skaffold [codebase](https://github.com/GoogleContainerTools/skaffold).*
+
+## Config schema
+
+We introduce an `ArtifactDependency` slice within `Artifact` in `config.go`
+```go
+type ArtifactDependency struct {
+	ImageName string `yaml:"image" yamltags:"required"`
+	Alias string `yaml:"alias,omitempty"`
+}
+```
+This allows us to define a build stanza like below where image `leeroy-app` requires image `simple-go-app`:
+```yaml
+build:
+ artifacts:
+   - image: simple-go-app
+   - image: leeroy-app
+     requires:
+       - image: simple-go-app
+         alias: BASE
+```
+Alias is a token that will be replaced with the image reference in the builder definition files. For example, the `docker` builder will use the alias as a build argument key like `BASE=simple-go-app:<tag>@sha:<sha>` passed as a `--build-arg` when building `leeroy-app`. If no value is provided for `alias` then it defaults to the value of `image`.
+
+## Config validation
+
+We add two new validations to the [validation](https://github.com/GoogleContainerTools/skaffold/blob/10275c66a142719897894308b9e566953712a0fe/pkg/skaffold/schema/validation/validation.go#L37) package after the introduction of artifact dependencies:
+- Cyclic references among artifacts.
+  - We cannot have image `A` depend on image `B` depend on image `C` depend on image `A`.
+  - We run a simple depth first search cycle detection algorithm treating our `Artifact` slice like a directed graph- image `A` depending on image `B` implies a directed edge from `A` to `B`.
+- Unique artifact aliases.
```

Comment:
Few things to consider, 
1. Can a base artifact be deployed?
- Is yes, then a base artifact can have sync rules and the would directly sync in the deployed container.
- if not, 
   - What happens to the sync rules defined for the artifact?
       a. Should they be combined with the sync rules defined in the dependency artifact?
            - If yes, what is the conflict resolution if they overlap.

In my opinion, a base artifact most of the times is a shared library code and not a binary. 
We can assume #1 and disallow any sync rules defined for a base artifact.
Add a section of assumptions and list this into assumptions and see what users have use cases where the deploy a base image as a container or they want to inherit sync rules from base image.
WDYT?

2. Since, this "alias" syntax is only applicable for docker builder, `requires` should be allowed only on `DockerArtifact`.

Reply:
> 1.
Yes base artifact can be deployed and it can define its own sync rules. When considering it as a dependency however we are just concerned with the image that it builds and ignore any sync rules. I don't think that there's a reliable way to merge base sync rules since we don't really know how the base image is gonna be used in the dependent artifact's build. WDYT?
cc. @briandealwis 
> 2.
I added that to the document scope. We can add a validation to check this and remove it later as we make progress with other builders.

Reply:
https://github.com/GoogleContainerTools/skaffold/issues/3247 and https://github.com/GoogleContainerTools/skaffold/issues/889 are two reported instances. I don't think there are any reported cases involving BASE artifacts that have sync definitions. 

>Consider a case where shared are assets are defined in a base image. Ideally, sync rules for this asset would be part of the artifact. If they are ignored, they would end up duplicating in all images where they are used.

This example definitely helps me understand the confusion that you're talking about. Maybe I'm wrong about this: "don't think that there's a reliable way to merge base sync rules"

We can probably, with some effort parsing the Dockerfile, figure out if the derived artifact is copying selective files like in a multistage build or just using it as a base image and try to create matching sync rules in the derived artifact? Anyway seems like a good followup candidate; we can start off by forcing a rebuild and ignore being able to just sync the files. If there are enough objections to that approach we can try to optimize later? WDYT?

Reply:
A very thought-provoking question, @tejal29.  Syncing and base images has deep implications, especially since we now provide a SKAFFOLD_RUN_MODE that a Dockerfile could use to alter its behaviour, such as to run a file-watcher to process sync'd files.  Those file-watchers may not be invoked in child images.

For example, consider a base image that sets up a watcher to compiles a set of .java files to .jar files (e.g., `gradle --continuous`).  This image is used by a set of child images that run different jars with different entrypoints.  If we inherit sync rules from the base, then there's a strong possibility that the sync'd files are ignored.

It seems to me that images used as dependencies may need to be built with SKAFFOLD_RUN_MODE=build, and any attending changes to build-flags.

I'll note that syncing is an optimization — we should always be able to fall-back to rebuilding images.

Reply:
trying to merge sync rules from a base image seems like it will be difficult to get right in practice - there are lots of scenarios I can imagine where sync rules from the base overwrite the pre-existing sync rules from a dependent image in a way that's unexpected for the user.

for the first implementation I think we should just ignore sync rules from a base image when it isn't only being deployed on its own. we could consider allowing annotations on the base image to tell skaffold to honor its sync rules, but this should be a workaround and not the default.